src/lib/socket.c

```diff
@@ -2362,46 +2397,53 @@ static long bio_ctrl(BIO *bio, int cmd, long udata_i, void *udata_p) {
 } /* bio_ctrl() */
 
 static int bio_create(BIO *bio) {
-	bio->init = 0;
-	bio->shutdown = 0;
-	bio->num = 0;
-	bio->ptr = NULL;
+	BIO_set_init(bio, 0);
+	BIO_set_shutdown(bio, 0);
+	BIO_set_data(bio, NULL);
 
 	return 1;
 } /* bio_create() */
 
 static int bio_destroy(BIO *bio) {
-	bio->init = 0;
-	bio->shutdown = 0;
-	bio->num = 0;
-	bio->ptr = NULL;
+	BIO_set_init(bio, 0);
+	BIO_set_shutdown(bio, 0);
+	BIO_set_data(bio, NULL);
 
 	return 1;
 } /* bio_destroy() */
 
-static BIO_METHOD bio_methods = {
-	BIO_TYPE_SOURCE_SINK,
-	"struct socket*",
-	bio_write,
-	bio_read,
-	bio_puts,
-	NULL,
-	bio_ctrl,
-	bio_create,
-	bio_destroy,
-	NULL,
-};
+static BIO_METHOD *bio_methods;
 
 static BIO *so_newbio(struct socket *so, int *error) {
 	BIO *bio;
 
-	if (!(bio = BIO_new(&bio_methods))) {
+	if (bio_methods == NULL) {
+		int type = BIO_get_new_index();
+		if (type == -1) {
+			*error = SO_EOPENSSL;
+			return NULL;
+		}
+		type |= BIO_TYPE_SOURCE_SINK|BIO_TYPE_DESCRIPTOR;
+		bio_methods = BIO_meth_new(type, "struct socket*");
```

Comment:
Not if the code were statically compiled, like if statically linking modules into an existing binary. The Debian Lua policy requires being able to produce liblua-cqueues.a. cqueues doesn't do that now (at least, not out-of-the-box), but I think it's a good idea. It would be awesome to be able to build and distribute Lua programs with everything statically included.